graphql/src/main/java/com/linecorp/armeria/server/graphql/GraphqlWSSubProtocol.java

```diff
@@ -0,0 +1,353 @@
+                    writeConnectionAck(out);
+                    break;
+                case "ping":
+                    writePong(out);
+                    break;
+                case "pong":
+                    break;
+                case "subscribe":
+                    ensureInitiated();
+                    id = toStringFromJson("id", eventMap.get("id"));
+                    requireNonNull(id, "id");
+                    final Map<String, Object> payload = toMapFromJson(eventMap.get("payload"));
+                    try {
+                        if (graphqlSubscriptions.containsKey(id)) {
+                            throw new IllegalArgumentException(
+                                    "Subscription with id " + id + " already exists");
+                        }
+                        final String operationName = toStringFromJson("operationName",
+                                                                      payload.get("operationName"));
+                        final String query = toStringFromJson("query", payload.get("query"));
+                        final Map<String, Object> variables = toMapFromJson(payload.get("variables"));
+                        final Map<String, Object> extensions = toMapFromJson(payload.get("extensions"));
+
+                        final ExecutionInput.Builder executionInput =
+                                ExecutionInput.newExecutionInput()
+                                              .graphQLContext(upgradeCtx)
+                                              .graphQLContext(connectionCtx)
+                                              .query(query)
+                                              .variables(variables)
+                                              .operationName(operationName)
+                                              .extensions(extensions);
+
+                        final CompletableFuture<ExecutionResult> future =
+                                graphqlExecutor.executeGraphql(ctx, executionInput);
+
+                        future.handleAsync(((executionResult, throwable) -> {
+                            handleExecutionResult(out, id, executionResult, throwable);
+                            return null;
+                        }), ctx.eventLoop());
+                    } catch (Exception e) {
+                        logger.debug("Error handling subscription", e);
+                        GraphqlWSSubProtocol.this.writeError(out, id, e);
+                        return;
+                    }
+                    break;
+                case "complete":
+                    ensureInitiated();
+                    // Read id and remove that subscription
+                    id = toStringFromJson("id", eventMap.get("id"));
+                    requireNonNull(id, "id");
+                    final ExecutionResultSubscriber s = graphqlSubscriptions.remove(id);
+                    if (s != null) {
+                        s.setCompleted();
```

Comment:
If a subscription isn't received yet (`ExecutionResultSubscriber.subscription == null`) , I guess a stream may not be completed.

What do you think of modifying `ExecutionResultSubscriber` so that it can subscribe only once?
i.e.
ExecutionResultSubscriber.java
```
@Override
public void onSubscribe(Subscription s) {
    if (subscription != null) {
        s.cancel();
        return;
    }
...
    public void setCompleted() {
        if (subscription == null) {
            subscription = NoopSubscription.get();
        }
        subscription.cancel();
    }
```
This way, a late subscription will still be cancelled